backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/questionsubmission/QuestionSubmissionService.java

```diff
@@ -0,0 +1,247 @@
+        checkIfConsistentReview(reviewDto);
+
+        QuestionSubmission questionSubmission = getQuestionSubmission(reviewDto.getQuestionSubmissionId());
+
+        User user = getUser(reviewDto.getUserId());
+
+        Review review = new Review(user, questionSubmission, reviewDto);
+
+        updateQuestionStatus(review.getStatus(), questionSubmission.getQuestion().getId());
+
+        reviewRepository.save(review);
+        return new ReviewDto(review);
+    }
+
+    @Retryable(value = {SQLException.class}, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public QuestionSubmissionDto updateQuestionSubmission(Integer questionSubmissionId, QuestionSubmissionDto questionSubmissionDto) {
+        QuestionSubmission questionSubmission = getQuestionSubmission(questionSubmissionId);
+
+        if (questionSubmission.getQuestion().getStatus() != Question.Status.IN_REVISION) {
+            throw new TutorException(CANNOT_EDIT_REVIEWED_QUESTION);
+        }
+
+        this.questionService.updateQuestion(questionSubmission.getQuestion().getId(), questionSubmissionDto.getQuestion());
+        return new QuestionSubmissionDto(questionSubmission);
+    }
+
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public void updateQuestionSubmissionTopics(Integer questionSubmissionId, TopicDto[] topics) {
+        Integer questionId = questionSubmissionRepository.findQuestionIdByQuestionSubmissionId(questionSubmissionId).orElse(null);
+
+        questionService.updateQuestionTopics(questionId, topics);
+    }
+
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public void removeSubmittedQuestion(Integer questionSubmissionId) {
+        QuestionSubmission questionSubmission = questionSubmissionRepository.findById(questionSubmissionId).orElseThrow(() -> new TutorException(QUESTION_SUBMISSION_NOT_FOUND, questionSubmissionId));
+        Question question = questionRepository.findById(questionSubmission.getQuestion().getId()).orElseThrow(() -> new TutorException(QUESTION_NOT_FOUND, questionSubmission.getQuestion().getId()));
+
+        if (!questionSubmission.getReviews().isEmpty() || !question.getStatus().equals(Question.Status.IN_REVISION)) {
+            throw new TutorException(CANNOT_DELETE_REVIEWED_QUESTION);
+        }
+
+        deleteQuestionSubmission(questionSubmission);
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public void deleteQuestionSubmission(QuestionSubmission questionSubmission) {
+        questionSubmission.remove();
+        questionSubmission.getReviews().forEach(review -> reviewRepository.delete(review));
```

Comment:
if we have a cascade from QuestionSubmission to Review, it would be enough to do review.remove(), which should be inside questionSubmission.remove()